icu4j/main/classes/core/src/com/ibm/icu/impl/units/ConversionRates.java

```diff
@@ -0,0 +1,234 @@
+        MeasureUnit.SIPrefix siPrefix = singleUnit.getSiPrefix();
+        UnitConverter.Factor result = UnitConverter.Factor.processFactor(mapToConversionRate.get(singleUnit.getSimpleUnit()).getConversionRate());
+
+        return result.applySiPrefix(siPrefix).power(power); // NOTE: you must apply the SI prefixes before the power.
+    }
+
+    public UnitConverter.Factor getFactorToBase(MeasureUnitImpl measureUnit) {
+        UnitConverter.Factor result = new UnitConverter.Factor();
+        for (SingleUnitImpl singleUnit :
+                measureUnit.getSingleUnits()) {
+            result = result.multiply(getFactorToBase(singleUnit));
+        }
+
+        return result;
+    }
+
+    protected BigDecimal getOffset(MeasureUnitImpl source, MeasureUnitImpl target, UnitConverter.Factor
+            sourceToBase, UnitConverter.Factor targetToBase, UnitConverter.Convertibility convertibility) {
+        if (convertibility != UnitConverter.Convertibility.CONVERTIBLE) return BigDecimal.valueOf(0);
+        if (!(checkSimpleUnit(source) && checkSimpleUnit(target))) return BigDecimal.valueOf(0);
+
+        String sourceSimpleIdentifier = source.getSingleUnits().get(0).getSimpleUnit();
+        String targetSimpleIdentifier = target.getSingleUnits().get(0).getSimpleUnit();
+
+        BigDecimal sourceOffset = this.mapToConversionRate.get(sourceSimpleIdentifier).getOffset();
+        BigDecimal targetOffset = this.mapToConversionRate.get(targetSimpleIdentifier).getOffset();
+        return sourceOffset
+                .subtract(targetOffset)
+                .divide(targetToBase.getConversionRate(), MathContext.DECIMAL128);
+
+
+    }
+
+    public MeasureUnitImpl getBasicMeasureUnitImplWithoutSIPrefix(MeasureUnitImpl measureUnit) {
+        ArrayList<SingleUnitImpl> baseUnits = this.getBasicUnitsWithoutSIPrefix(measureUnit);
+
+        MeasureUnitImpl result = new MeasureUnitImpl();
+        for (SingleUnitImpl baseUnit :
+                baseUnits) {
+            result.appendSingleUnit(baseUnit);
+        }
+
+        return result;
+    }
+
+    public ArrayList<SingleUnitImpl> getBasicUnitsWithoutSIPrefix(MeasureUnitImpl measureUnitImpl) {
+        ArrayList<SingleUnitImpl> result = new ArrayList<>();
+        ArrayList<SingleUnitImpl> singleUnits = measureUnitImpl.getSingleUnits();
+        for (SingleUnitImpl singleUnit :
+                singleUnits) {
+            result.addAll(getBasicUnitsWithoutSIPrefix(singleUnit));
+        }
+
+        return result;
+    }
+
+    /**
+     * @param singleUnit
+     * @return The bese units in the `SingleUnitImpl` with applying the dimensionality only and not the SI prefix.
+     * <p>
+     * NOTE:
+     * This method is helpful when checking the convertibility because no need to check convertibility.
+     */
+    public ArrayList<SingleUnitImpl> getBasicUnitsWithoutSIPrefix(SingleUnitImpl singleUnit) {
+        String target = mapToConversionRate.get(singleUnit.getSimpleUnit()).getTarget();
```

Comment:
This getTarget() might return "kilogram", which does have an SI prefix. Hence the C++ names can be confusing.

The concept of an "SI Base Unit" is a well-known one, see https://en.wikipedia.org/wiki/Base_unit_(measurement) - so some thought went into "extractCompoundBaseUnit". This doens't mean it's the best name ever, but it does mean it can be worthwhile to be mindful about the reasons you choose to change the name.

And names do matter: I apologise if I've at times been too nitpicky about names in previous pull requests, I'm trying to find the right balance. Hwoever I think the fact that you're using different names in this PR than we have in C++, even for code that you wrote yourself, is proof that the C++ names may have benefited from some more thought.

Reply:
Done!